File: skeleton/src/main/java/me/shkschneider/skeleton/helpers/IntentHelper.java
package me.shkschneider.skeleton.helpers;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.provider.Settings;
import android.support.annotation.NonNull;

import me.shkschneider.skeleton.SkeletonApplication;
import me.shkschneider.skeleton.java.StringHelper;
import me.shkschneider.skeleton.network.NetworkHelper;

import java.util.List;

public class IntentHelper {

    public static final int HOME_FLAGS = Intent.FLAG_ACTIVITY_CLEAR_TOP
            | Intent.FLAG_ACTIVITY_NEW_TASK
            | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT;

    public static final String BROADCAST_TIME_TICK = Intent.ACTION_TIME_TICK;
    public static final String BROADCAST_TIME_CHANGED = Intent.ACTION_TIME_CHANGED;
    public static final String BROADCAST_TIMEZONE_CHANGED = Intent.ACTION_TIMEZONE_CHANGED;
    public static final String BROADCAST_BOOT_COMPLETED = Intent.ACTION_BOOT_COMPLETED;
    public static final String BROADCAST_PACKAGE_ADDED = Intent.ACTION_PACKAGE_ADDED;
    public static final String BROADCAST_PACKAGE_CHANGED = Intent.ACTION_PACKAGE_CHANGED;
    public static final String BROADCAST_PACKAGE_REMOVED = Intent.ACTION_PACKAGE_REMOVED;
    public static final String BROADCAST_PACKAGE_RESTARTED = Intent.ACTION_PACKAGE_RESTARTED;
    public static final String BROADCAST_PACKAGE_DATA_CLEARED = Intent.ACTION_PACKAGE_DATA_CLEARED;
    public static final String BROADCAST_UID_REMOVED = Intent.ACTION_UID_REMOVED;
    public static final String BROADCAST_BATTERY_CHANGED = Intent.ACTION_BATTERY_CHANGED;
    public static final String BROADCAST_POWER_CONNECTED = Intent.ACTION_POWER_CONNECTED;
    public static final String BROADCAST_POWER_DISCONNECTED = Intent.ACTION_POWER_DISCONNECTED;
    public static final String BROADCAST_SHUTDOWN = Intent.ACTION_SHUTDOWN;

    private static final int REQUEST_CODE_CAMERA = 111;
    private static final int REQUEST_CODE_GALLERY = 222;

    public static boolean canHandle(@NonNull final Intent intent) {
        final PackageManager packageManager = SkeletonApplication.CONTEXT.getPackageManager();
        if (packageManager == null) {
            LogHelper.warning("PackageManager was NULL");
            return false;
        }

        List<ResolveInfo> resolveInfos = packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
        return (resolveInfos.size() > 0);
    }

    public static boolean home(@NonNull final Activity activity, @NonNull final Class cls) {
        Intent intent = new Intent(activity, cls);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
        activity.startActivity(intent);
        return true;
    }

    public static boolean web(@NonNull final Activity activity, @NonNull final String url) {
        if (! NetworkHelper.validUrl(url)) {
            LogHelper.warning("Url was invalid");
            return false;
        }

        activity.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
        return true;
    }

    public static boolean market(@NonNull final Activity activity, @NonNull final String pkg) {
        final Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse("market://details?androidId=" + pkg));
        activity.startActivity(intent);
        return true;
    }

    public static boolean market(@NonNull final Activity activity) {
        final Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse("market://details?androidId=" + ApplicationHelper.packageName()));
        activity.startActivity(intent);
        return true;
    }

    public static boolean email(@NonNull final Activity activity, @NonNull final String[] to, final String subject, final String text) {
        final Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("plain/text");
        intent.putExtra(Intent.EXTRA_EMAIL, to);
        if (! StringHelper.nullOrEmpty(subject)) {
            intent.putExtra(Intent.EXTRA_SUBJECT, subject);
        }
        if (! StringHelper.nullOrEmpty(text)) {
            intent.putExtra(Intent.EXTRA_TEXT, text);
        }
        activity.startActivity(Intent.createChooser(intent, "Send An Email"));
        return true;
    }

    public static boolean image(@NonNull final Activity activity, @NonNull final Uri uri) {
        final Intent intent = new Intent();
        intent.setAction(Intent.ACTION_VIEW);
        intent.setDataAndType(uri, "image/*");
        activity.startActivity(intent);
        return true;
    }

    public static boolean camera(@NonNull final Activity activity) {
        if (! FeaturesHelper.feature(FeaturesHelper.CAMERA)) {
            LogHelper.warning("Camera was unavailable");
            return false;
        }

        final Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (! canHandle(intent)) {
            LogHelper.warning("Cannot handle Intent");
            return false;
        }

        activity.startActivityForResult(intent, REQUEST_CODE_CAMERA);
        return true;
    }

    public static boolean gallery(@NonNull final Activity activity) {
        final Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setType("image/*");
        activity.startActivityForResult(intent, REQUEST_CODE_GALLERY);
        return true;
    }

    public static boolean settings(@NonNull final Activity activity) {
        activity.startActivity(new Intent(Settings.ACTION_SETTINGS));
        return true;
    }

    public static Bitmap onActivityResult(final int requestCode, final int resultCode, @NonNull final Intent intent) {
        if (resultCode != Activity.RESULT_OK) {
            LogHelper.debug("ResultCode was not OK");
            return null;
        }

        if (requestCode == REQUEST_CODE_CAMERA) {
            final Bundle bundle = intent.getExtras();
            if (bundle == null) {
                LogHelper.warning("Bundle was NULL");
                return null;
            }

            return (Bitmap) bundle.get("data");
        }
        else if (requestCode == REQUEST_CODE_GALLERY) {
            final Uri uri = intent.getData();
            if (uri == null) {
                LogHelper.warning("Uri was NULL");
                return null;
            }

            return BitmapHelper.decodeUri(uri);
        }
        return null;
    }

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/content/IntentHelper.kt
package me.shkschneider.skeleton.android.content

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.media.RingtoneManager
import android.net.Uri
import android.provider.ContactsContract
import android.provider.MediaStore
import android.provider.Settings
import me.shkschneider.skeleton.android.app.ApplicationHelper
import me.shkschneider.skeleton.android.graphics.BitmapHelper
import me.shkschneider.skeleton.android.os.LocaleHelper
import me.shkschneider.skeleton.kotlin.data.MimeTypes
import me.shkschneider.skeleton.android.network.UrlHelper
import me.shkschneider.skeleton.android.os.AndroidHelper
import me.shkschneider.skeleton.android.provider.ContextProvider
import me.shkschneider.skeleton.android.provider.AndroidFeature
import me.shkschneider.skeleton.kotlin.log.Logger
import java.io.File

// <http://developer.android.com/reference/android/content/Intent.html>
object IntentHelper {

    const val FLAGS_HOME = (Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
    const val FLAGS_CLEAR = (Intent.FLAG_ACTIVITY_REORDER_TO_FRONT or Intent.FLAG_ACTIVITY_CLEAR_TOP)

    object GooglePlay {

        fun application(packageName: String): Intent {
            return external(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$packageName")))
        }

        fun publisher(pub: String): Intent {
            return external(Intent(Intent.ACTION_VIEW).apply {
                data = Uri.parse("market://search?pub:$pub")
            })
        }

        fun search(q: String): Intent {
            return external(Intent(Intent.ACTION_VIEW).apply {
                data = Uri.parse("market://search?q=$q")
            })
        }

    }

    private const val REQUEST_CODE_CAMERA = 111
    private const val REQUEST_CODE_GALLERY = 222
    private const val REQUEST_CODE_RINGTONE = 333
    private const val REQUEST_CODE_PERMISSIONS = AndroidHelper.ANDROID_6

    fun main(): Intent? {
        ApplicationHelper.packageManager.getLaunchIntentForPackage(ApplicationHelper.packageName)?.let { intent ->
            return Intent.makeMainActivity(intent.component).addFlags(FLAGS_HOME)
        }
        Logger.warning("Intent was NULL")
        return null
    }

    fun restart(): Intent? {
        ApplicationHelper.packageManager.getLaunchIntentForPackage(ApplicationHelper.packageName)?.let { intent ->
            return Intent.makeRestartActivityTask(intent.component).addFlags(FLAGS_CLEAR)
        }
        Logger.warning("Intent was NULL")
        return null
    }

    fun view(uri: Uri): Intent =
        external(Intent(Intent.ACTION_VIEW, uri))

    fun web(url: String): Intent? {
        if (!UrlHelper.valid(url)) {
            Logger.warning("Url was invalid")
            return null
        }
        return external(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    fun share(subject: String?, text: String?): Intent =
        Intent.createChooser(Intent(Intent.ACTION_SEND).apply {
            type = MimeTypes.TEXT_PLAIN
            // Email: data = Uri.parse("mailto:"))
            //        putExtra(Intent.EXTRA_EMAIL, new String[] { ... })
            if (!subject.isNullOrBlank()) {
                putExtra(Intent.EXTRA_SUBJECT, subject)
            }
            if (!text.isNullOrBlank()) {
                putExtra(Intent.EXTRA_TEXT, text)
            }
        }, null)

    fun directions(fromLatitude: Long, fromLongitude: Long, toLatitude: Long, toLongitude: Long): Intent =
        external(Intent(Intent.ACTION_VIEW,
                Uri.parse(String.format(LocaleHelper.Device.locale(),
                        "http://maps.google.com/maps?saddr=%s,%s&daddr=%s,%s",
                        fromLatitude, fromLongitude,
                        toLatitude, toLongitude))))

    fun ringtone(existingUri: String?): Intent =
        Intent(RingtoneManager.ACTION_RINGTONE_PICKER).apply {
            putExtra(RingtoneManager.EXTRA_RINGTONE_TYPE, RingtoneManager.TYPE_NOTIFICATION)
            putExtra(RingtoneManager.EXTRA_RINGTONE_EXISTING_URI, Uri.parse(existingUri.orEmpty()))
        }

    fun applicationSettings(): Intent =
        external(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
            data = Uri.parse("package:" + ApplicationHelper.packageName)
        })

    fun systemSettings(): Intent =
        external(Intent(Settings.ACTION_SETTINGS))

    fun text(uri: Uri): Intent =
        external(Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, MimeTypes.TEXT_PLAIN)
        })

    fun audio(uri: Uri): Intent =
        external(Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, MimeTypes.AUDIO)
        })

    fun video(uri: Uri): Intent =
        external(Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, MimeTypes.VIDEO)
        })

    fun picture(uri: Uri): Intent =
        external(Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, MimeTypes.IMAGE)
        })

    fun gallery(): Intent =
        external(Intent(Intent.ACTION_PICK).apply {
            type = MimeTypes.IMAGE
        })

    fun camera(file: File): Intent? {
        if (!AndroidFeature.Camera.isAvailable) {
            Logger.warning("Camera was unavailable")
            return null
        }
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
                .putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(file))
                .putExtra(MediaStore.EXTRA_SHOW_ACTION_ICONS, true)
        if (!canHandle(intent)) {
            Logger.warning("Cannot handle Intent")
            return null
        }
        return external(intent)
    }

    fun file(): Intent =
        external(Intent(Intent.ACTION_GET_CONTENT).apply {
            type = MimeTypes.FILE
        })

    fun dial(phone: String): Intent =
        external(Intent(Intent.ACTION_DIAL).apply {
            data = Uri.parse("tel:$phone")
        })

    fun call(phone: String): Intent =
        external(Intent(Intent.ACTION_CALL).apply {
            data = Uri.parse("tel:$phone")
        })

    fun sms(phone: String, message: String): Intent =
        external(Intent(Intent.ACTION_SENDTO).apply {
            data = Uri.parse("smsto:$phone")
            putExtra("sms_body", message)
        })

    fun contact(): Intent =
        external(Intent(Intent.ACTION_PICK, Uri.parse("content://com.android.contacts/contacts")).apply {
            type = ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE
        })

    fun canHandle(intent: Intent): Boolean {
        // return (intent.resolveActivity(ApplicationHelper.packageManager()) != null)
        val resolveInfos = ApplicationHelper.packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)
        return resolveInfos.size > 0
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent? = null): Any? {
        if (resultCode != Activity.RESULT_OK) {
            Logger.debug("ResultCode was not OK")
            return null
        }
        when (requestCode) {
            REQUEST_CODE_CAMERA -> {
                intent?.extras?.get("data")?.let { data ->
                    return data
                }
                Logger.warning("Intent's data was NULL")
                return null
            }
            REQUEST_CODE_GALLERY -> {
                intent?.data?.let { uri ->
                    return BitmapHelper.decodeUri(uri)
                }
                Logger.warning("Intent's data was NULL")
                return null
            }
            REQUEST_CODE_RINGTONE -> {
                intent?.getParcelableExtra<Uri>(RingtoneManager.EXTRA_RINGTONE_PICKED_URI)?.let { uri ->
                    return RingtoneManager.getRingtone(ContextProvider.applicationContext(), uri)
                }
                Logger.warning("Intent's extra was NULL")
                return null
            }
            else -> return null
        }
    }

    // <http://developer.android.com/training/implementing-navigation/descendant.html#external-activities>
    @Suppress("DEPRECATION")
    private fun external(intent: Intent): Intent =
        with(intent) {
            addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
        }

}